pkg/util/atomicstats/reporter.go

```diff
@@ -0,0 +1,461 @@
+	"regexp"
+	"strings"
+	"time"
+	"unsafe"
+
+	"go.uber.org/atomic"
+)
+
+const statsTag = "stats"
+
+// Reporter supports reporting on a specific struct type, containing
+// precomputed values to ensure subsequent operations are faster.
+//
+// Typically a single Reporter is created at startup for each stats struct.
+//
+// Such structs should tag fields to be included in the stats with `stats:""`.
+// Stats fields can be of any of the following types:
+//
+//     int
+//     int8
+//     int16
+//     int32
+//     int64
+//     uint
+//     uint8
+//     uint16
+//     uint32
+//     uint64
+//     uintptr
+//     go.uber.org/atomic.Bool
+//     go.uber.org/atomic.Duration
+//     go.uber.org/atomic.Error				// atomic swap not supported
+//     go.uber.org/atomic.Float64
+//     go.uber.org/atomic.Int32
+//     go.uber.org/atomic.Int64
+//     go.uber.org/atomic.String			// atomic swap not supported
+//     go.uber.org/atomic.Time				// atomic swap not supported
+//     go.uber.org/atomic.Uint32
+//     go.uber.org/atomic.Uint64
+//     go.uber.org/atomic.Uintptr
+//     go.uber.org/atomic.UnsafePointer
+//     go.uber.org/atomic.Value
+type Reporter struct {
+	// structType is the type of value this Reporter reports on
+	structType reflect.Type
+
+	// fields maps field names to getters and setters for that field
+	fields map[string]field
+}
+
+type field struct {
+	// idx is the index of the field in the struct
+	idx int
+
+	// report will get the value from the field
+	report func(f reflect.Value) interface{}
+
+	// reset will get the value from the field and reset it to its
+	// zero value.
+	reset func(f reflect.Value) interface{}
+}
+
+// NewReporter creates a new Reporter to represent the type of the argument.
+//
+// Typically the argument is a nil pointer of the correct type:
+//
+//     type foostats struct { .. }
+//     statsReporter := atomicstats.NewReporter((*foostats)(nil))
+func NewReporter(v interface{}) Reporter {
+	ptrType := reflect.TypeOf(v)
+
+	if ptrType.Kind() != reflect.Ptr || ptrType.Elem().Kind() != reflect.Struct {
+		// errors here are programming (type) errors that cannot be caught at compile time,
+		// and thus are fatal
+		panic("NewReporter expects a pointer to a struct")
+	}
+
+	structType := ptrType.Elem()
+
+	reporter := Reporter{
+		structType: structType,
+		fields:     map[string]field{},
+	}
+
+	for idx := 0; idx < structType.NumField(); idx++ {
+		fieldType := structType.Field(idx)
+		if _, ok := fieldType.Tag.Lookup(statsTag); ok {
+			fld, err := getFieldFor(fieldType.Type)
+			if err != nil {
+				panic(err.Error())
+			}
+
+			fld.idx = idx
+			reporter.fields[toSnakeCase(fieldType.Name)] = fld
+		}
+	}
+
+	return reporter
+}
+
+// Report returns a `map` representation of the stats in the given value.
+// All map keys are converted to snake_case.
+//
+// The value _must_ have the same type as the value passed to NewReporter.
+func (r *Reporter) Report(v interface{}) map[string]interface{} {
```

Comment:
We can't stash `v` in the `Reporter` instance?

Reply:
I changed that arrangement because the reporter isn't specific to the _instance_ of the reported-on struct, but to the type.  AIUI in pkg/network there's really only one instance of each type, so the distinction isn't too important, but I want to be able to use this for situations where there are lots of instances (e.g., one per check)

Reply:
Makes sense